### ISSUE TITLE
Refactor ruby/streaming integration surface

### ISSUE THREAD

**Issue by ThisIsMissEm** [mastodon/mastodon#28010]
### Pitch

At present, the code for interacting with the streaming server is scattered throughout the codebase as direct `redis.publish` calls, which makes it hard to evolve and test.

In order to improve the integration, I'd like to introduce a `Streaming` module/class that has methods for distributing messages to streaming. It'd have an API centered around the event being sent to streaming, e.g.,

```
Streaming.publish_timeline(timeline, event, status) # where event is 'update', 'status.update', 'delete', etc
Streaming.publish_conversation(account_id, conversation)
Streaming.revoke_access_token(token_id)
Streaming.update_custom_filters(account_id)
Streaming.update_filters(account_id, type) # type is blocks, blocking, mutes, muting
```

This would also handle all the serialization and sending to different topics as necessary, allowing us to ensure consistent serialisation of all messages going over redis pubsub.

### Motivation

This will help improve the testability of the codebase, and normalise how we interact from Ruby with the streaming server.

**Comment by ThisIsMissEm** [comment]
cc @renchap this is what I was trying to mention in the past.

**Comment by ThisIsMissEm** [comment]
The `Streaming.update_filters` is particularly needed, as it'd allow for us to move the database dependent code from the main message handling loop in the streaming server (as it'd be pushing changes to blocks/mutes into a system channel)

As an aside, we only really need a single system channel, not one per access token / account (it's incredibly difficult to actually shared redis pubsub as far as I know, and the benefits are likely to be minimal)

**Comment by renchap** [comment]
cc @ClearlyClaire

**Comment by ThisIsMissEm** [comment]
> The `Streaming.update_filters` is particularly needed, as it'd allow for us to move the database dependent code from the main message handling loop in the streaming server (as it'd be pushing changes to blocks/mutes into a system channel)
> 
> As an aside, we only really need a single system channel, not one per access token / account (it's incredibly difficult to actually shared redis pubsub as far as I know, and the benefits are likely to be minimal)

I did some early work on this in https://github.com/mastodon/mastodon/pull/28033/

**Comment by ThisIsMissEm** [comment]
Another change related here is the one for getting Streaming to receive the `searchable_text` as defined in `StatusSearchConcern` from Ruby, which allows us to completely remove JSDOM from the streaming codebase (which will make the service a heap lighter and more performant)

**Comment by mjankowski** [comment]
This is a good idea.

There are many direct accesses of the `redis` object across the whole app where I know that I'm really looking at the implementation detail of a higher level concept, and I need to pause a moment and contemplate which one.

A fun side effect of doing this would (I suspect?...) be to wind up with different higher level conceptual classes for the different types of stuff we use redis as the data store for. Starting with basically any one of the mentioned streaming concepts here and seeing how it evolves from there seems reasonable.

**Comment by ThisIsMissEm** [comment]
Yup, I just need the time to do all this work. It's not been my highest priority, because I know it'll be slow to ship.
